71_RayTracingPipeline/app_resources/common.hlsl

```diff
@@ -104,16 +157,55 @@ struct RayLight
 
 #ifdef __HLSL_VERSION
 
-struct [raypayload] ColorPayload
+struct [raypayload] OcclusionPayload
+{
+    float32_t attenuation : read(caller) : write(caller, anyhit);
+};
+
+struct MaterialId
 {
-	float32_t3 hitValue : read(caller) : write(closesthit,miss);
-    uint32_t seed : read(closesthit,anyhit) : write(caller);
+    const static uint32_t PROCEDURAL_FLAG = (1 << 31);
+    const static uint32_t PROCEDURAL_MASK = ~PROCEDURAL_FLAG;
+
+    uint32_t data;
+
+    static MaterialId createProcedural(uint32_t index)
+    {
+        MaterialId id;
+        id.data = index | PROCEDURAL_FLAG;
+        return id;
+    }
+
+    static MaterialId createTriangle(uint32_t index)
+    {
+        MaterialId id;
+        id.data = index;
+        return id;
+    }
+
+    uint32_t getMaterialIndex()
+    {
+        return data & PROCEDURAL_MASK;
+    }
+
+    bool isHitProceduralGeom()
+    {
+        return data & PROCEDURAL_FLAG;
+    }
 };
 
-struct [raypayload] ShadowPayload
+struct [raypayload] PrimaryPayload
 {
-	bool isShadowed : read(caller) : write(caller,miss);
-    uint32_t seed : read(anyhit) : write(caller);
+    float32_t3 worldNormal : read(caller) : write(closesthit);
+    float32_t rayDistance : read(caller) : write(closesthit, miss);
```

Comment:
isn't that already available by the time your `trace` intrinsic returns? or is that only with the new SER extensions?

Reply:
yep SER extensions make ray state available after traceRay is called